### DISCUSSION TITLE
指令集架构、机器码与 Go 语言 · /golang-machinecode

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#541] [Announcements]
https://draveness.me/golang-machinecode

**Comment by two** [comment]
🆙

**Comment by Devying** [comment]
很想知道，楼主是做什么开发的？对底层这些东西了解的十分透传，这是我等凡人无法比的

**Comment by maodunWorld** [comment]
哥，问一句，对于AMD CPU和Intel CPU，Golang需要重新编译吗？

**Comment by draveness** [comment]
> 哥，问一句，对于AMD CPU和Intel CPU，Golang需要重新编译吗？

这要看它俩指令集一样不，然后你说的是 Go 语言本身还是 Go 的项目...

**Comment by maodunWorld** [comment]
我说的是Go Mod项目，比如说，我build一个go为main.exe，可以在AMD R5下运行，但是我部署的服务器是Intel上的，这个会有影响吗？

**Comment by draveness** [comment]
> 我说的是Go Mod项目，比如说，我build一个go为main.exe，可以在AMD R5下运行，但是我部署的服务器是Intel上的，这个会有影响吗？

这俩机器的指令集一样就不用重新编译，如果不一样的话，需要在构建的时候指定 OS 和 Arch

**Comment by konkona** [comment]
楼主太厉害了，越学越觉得自己的渺小。。。

**Comment by draveness** [comment]
> 楼主太厉害了，越学越觉得自己的渺小。。。

😂😂😂

**Comment by yuchanns** [comment]
```go
MOVQ (CX), CX
```
看`for range`那一节的时候试着输出了一下汇编的代码，但是看这个不太理解是什么意思，特来询问一下。是否可以理解为 硬件CX的值移动到伪CX上？看了[官方文档asm](https://golang.org/doc/asm#symbols)提到的伪寄存器，并没有提到CX也是伪的

**Comment by yuchanns** [comment]
> ```go
> MOVQ (CX), CX
> ```
> 
> 看`for range`那一节的时候试着输出了一下汇编的代码，但是看这个不太理解是什么意思，特来询问一下。是否可以理解为 硬件CX的值移动到伪CX上？看了[官方文档asm](https://golang.org/doc/asm#symbols)提到的伪寄存器，并没有提到CX也是伪的

哦，我似乎明白了，这应该是`CX =  *CX`的意思

**Comment by draveness** [comment]
> > ```go
> > MOVQ (CX), CX
> > ```
> > 
> > 
> > 看`for range`那一节的时候试着输出了一下汇编的代码，但是看这个不太理解是什么意思，特来询问一下。是否可以理解为 硬件CX的值移动到伪CX上？看了[官方文档asm](https://golang.org/doc/asm#symbols)提到的伪寄存器，并没有提到CX也是伪的
> 
> 哦，我似乎明白了，这应该是`CX = *CX`的意思

对的

**Comment by yuchanns** [comment]
```go
MOVQ    AX, (BX)(DX*8)
```
那这个是什么意思呢？主要是不太理解`*8`是什么意思

**Comment by draveness** [comment]
> @yuchanns
> ```go
> MOVQ    AX, (BX)(DX*8)
> ```
> 那这个是什么意思呢？主要是不太理解`*8`是什么意思

我也不能确定，看看编译前的源代码查查手册吧

**Comment by draveness** [comment]
> ```go
> MOVQ    AX, (BX)(DX*8)
> ```
> 
> 那这个是什么意思呢？主要是不太理解`*8`是什么意思

Addressing modes:

+ (DI)(BX*2): The location at address DI plus BX*2.
+ 64(DI)(BX*2): The location at address DI plus BX*2 plus 64. These modes accept only 1, 2, 4, and 8 as scale factors.

这是我刚查到的，在你刚发的链接里 https://golang.org/doc/asm#x86

**Comment by Peakchen** [comment]
linux机器码怎么获得的

**Comment by draveness** [comment]
> linux机器码怎么获得的

什么意思

**Comment by seapub** [comment]
作为一种常见的精简指令集处理器，amd 使用 4 个字节作为指令的固定长度，省略了判断指令的性能损失
根据上下文,我猜应该是 arm

**Comment by draveness** [comment]
> 作为一种常见的精简指令集处理器，amd 使用 4 个字节作为指令的固定长度，省略了判断指令的性能损失
> 根据上下文,我猜应该是 arm

确实是...

**Comment by yanjinbin** [comment]
清明假期 又发现了一块宝藏知识， 越学越觉得自己是真滴菜。
编译原理这一章节，学到了点关于Go编译的知识，
发现  编译原理  还是能涉及到很多纯理论和数据结构 算法的运用。
好奇的是，楼主做到这些？把 编译的几个核心点映射到go编译上，从哪里入手的呢？好奇

**Comment by draveness** [comment]
> 清明假期 又发现了一块宝藏知识， 越学越觉得自己是真滴菜。
> 编译原理这一章节，学到了点关于Go编译的知识，
> 发现 编译原理 还是能涉及到很多纯理论和数据结构 算法的运用。
> 好奇的是，楼主做到这些？把 编译的几个核心点映射到go编译上，从哪里入手的呢？好奇

之前学过编译原理，也写过 Parser

**Comment by sabragirl** [comment]
倒数第二个代码段中，这里可能是想用 hello.go 我猜，不过没啥影响
```
$ GOOS=linux GOARCH=amd64 go tool compile -S main.go  # hello.go
```

---

*2020-04-28 UPDATES: 已修复*

**Comment by soyacen** [comment]
大神需要马仔吗？

**Comment by draveness** [comment]
> 大神需要马仔吗？

你猜

**Comment by denghui** [comment]
复杂的机器指定对于编译器来说很难利用

这里有个typo，指定=指令？

---

*2020-07-31 UPDATES: 已修复*

**Comment by woofyzhao** [comment]
上一节里中间代码中被改写的内置函数和关键字的运行时函数最终是怎么进到目标程序里的？

**Comment by draveness** [comment]
> 上一节里中间代码中被改写的内置函数和关键字的运行时函数最终是怎么进到目标程序里的？

编译的时候源代码会被翻译成二进制

**Comment by woofyzhao** [comment]
> @draveness
> > 上一节里中间代码中被改写的内置函数和关键字的运行时函数最终是怎么进到目标程序里的？
> 
> 编译的时候源代码会被翻译成二进制

运行时函数不是事先编好的吗？我理解是不是比如AST chan接收操作结点被改写chanrecv1后就以纯**symbol**的形式带下去，包括生成ssa中间代码、过汇编器也一直只是个symbol，最后链接成elf的时候才去事先编好的运行时pkg里去找代码？

**Comment by draveness** [comment]
> > @draveness
> > > 上一节里中间代码中被改写的内置函数和关键字的运行时函数最终是怎么进到目标程序里的？
> > 
> > 
> > 编译的时候源代码会被翻译成二进制
> 
> 运行时函数不是事先编好的吗？我理解是不是比如AST chan接收操作结点被改写chanrecv1后就以纯**symbol**的形式带下去，包括生成ssa中间代码、过汇编器也一直只是个symbol，最后链接成elf的时候才去事先编好的运行时pkg里去找代码？

嗯呢，我理解的是你问的前半部分

**Comment by zkfopen** [comment]
但是在这里还是需要引入更多的指令集构知识。
修改为：
但是在这里还是需要引入更多的指令集架构知识。

---

*2021-03-15 UPDATES: 已修复*

**Comment by bugangongwei** [comment]
> @soyacen
> 大神需要马仔吗？

"马仔是一个汉语词语，读音为mǎ zǎi，一指马驹，二指方言，流氓恶霸的狗腿、帮手。"
 -- 百度百科

**Comment by draveness** [comment]
> @bugangongwei 
> > @soyacen
> > 大神需要马仔吗？
> 
> "马仔是一个汉语词语，读音为mǎ zǎi，一指马驹，二指方言，流氓恶霸的狗腿、帮手。"
> -- 百度百科

🤣

**Comment by sj82516** [comment]
太精彩了！ 
想請教樓主「cmd/compile/internal/ssa 主要负责对 SSA 中间代码进行降级、执行架构特定的优化和重写并生成 cmd/compile/internal/obj.Prog 指令」這可以說是產生出 assembly code 嗎？
主要是當初看到 Rob Pike 關於 Go Assembly 的演講 https://www.youtube.com/watch?v=KINIAgRpkDA ，提到 Go 有將 source code 先編譯成  Plan9 的匯編，接著在透過 obj tool 根據不同的 GOARCH 產生對應的機器碼
想請教樓主這樣的描述是否精確

謝謝

**Comment by iam153** [comment]
博主你好，我想问下，pp.Flushplist上面有一句注释
// Flush converts from pp to machine code.
这里的pp(Progs)能当做汇编看待吗？如果不能的话，在这个函数里从Progs转成汇编的是哪部分呢。谢谢博主。

**Comment by iam153** [comment]
花了一周时间把第二章似懂非懂地啃完了，现在居然还有点不舍的感觉。看博客评论区似乎大佬最近不太关注这块了，回复少了很多，希望大佬多出点文章。

**Comment by qqlcbb** [comment]
精彩

**Comment by draveness** [comment]
> 花了一周时间把第二章似懂非懂地啃完了，现在居然还有点不舍的感觉。看博客评论区似乎大佬最近不太关注这块了，回复少了很多，希望大佬多出点文章。

最近事情实在是太多了

**Comment by zyj123** [comment]
> @draveness
> > linux机器码怎么获得的
> 
> 什么意思

go build

**Comment by qiqiandjing** [comment]
总结一下看看对不对
源码->此法解析->语法解析->语法树->类型检验->语法树->中间代码->汇编->机器码 
不知过程是否对

**Comment by Juminiy** [comment]
編譯原理只講述了一部分，對於一個狀態機來講，是需要內存分配的策略，這個策略所屬的狀態機可以是組譯器，操作系統，web服務器，還是期待博主的內存分配策略和寄存器著色算法。
